daft/filesystem.py

```diff
@@ -29,14 +31,36 @@ def _get_fs_from_cache(protocol: str, io_config: IOConfig | None) -> pafs.FileSy
     """
     global _CACHED_FSES
 
-    return _CACHED_FSES.get((protocol, io_config))
+    if (protocol, io_config) in _CACHED_FSES:
+        fs, expiry = _CACHED_FSES[(protocol, io_config)]
+        print("found cached fs")
 
+        if expiry is None or expiry > datetime.now(timezone.utc):
+            print("cached fs not expired")
+            return fs
 
-def _put_fs_in_cache(protocol: str, fs: pafs.FileSystem, io_config: IOConfig | None) -> None:
+    return None
+
+
+def _put_fs_in_cache(protocol: str, fs: pafs.FileSystem, io_config: IOConfig | None, expiry: datetime | None) -> None:
     """Put pyarrow filesystem in cache under provided protocol."""
     global _CACHED_FSES
 
-    _CACHED_FSES[(protocol, io_config)] = fs
+    _CACHED_FSES[(protocol, io_config)] = (fs, expiry)
+
+
+def _get_s3_creds_from_provider_cached(provider: Callable[[], S3Credentials]) -> S3Credentials:
+    """Get S3 credentials from the cache if the current provider was already called and the creds have not expired."""
+    global _CACHED_S3_CREDS
```

Comment:
What about "always" running the provider instead of caching it? Global state is scary and bug-prone.

Doesn't `_CACHED_FSES` already handle caching for us, since this maps the IOConfig (which contains a provider) to the initialized fs?

Reply:
It does in part, but fs objects are still created every time `_resolve_paths_and_filesystem` is called as part of the call to `_validate_filesystem`, even though the cached one is ultimately returned. Because of that, the S3 credentials will retrieved again, potentially multiple times per filesystem resolution actually (`_validate_filesystem` is called once per path)

I am honestly not sure why we cache the fs but didn't want to change that behavior in this PR

Reply:
Ok... Another idea that might be a little cleaner -- could we wrap the `provider` callback into a cached class? This could help us avoid a global cache I think.

Something like this:

```python
class CachedProvider:
    def __init__(self, provider):
        self.expiry = ...
        self.provider = ...
        self.creds = None

    def __call__(self):
        ... # used cached if expiry is still good, otherwise re-trigger the provider

    def __reduce__(self):
        ... # for pickling, we can discard the cached credentials and just pickle the provider itself I think
```

Reply:
Hm, where would I store this cached class? I'm hesitant to overload `S3Config.credentials_provider` but unsure if there are other options

Reply:
Yeah I was thinking we store it in there as a little wrapper!

Reply:
Ended up going with a different solution actually, adding a caching wrapper on the Rust side. What do you think?

Reply:
Oh very cool!!